### PR TITLE
LOG-9749 cli introduces a new concept of log groups and nick names

### PR DESCRIPTION
**Pull request by pclune-r7**
@ilyabiryukov-r7 @pquinn-r7 
Howdy gents!
Looking for a review.
Thanks!

Commits on the work done that have been squashed:

- LOG-9749 Removal of query nicknames use
- LOG-9749 Introduce CLI Favorites, remove reference to Log Nicknames
- LOG-9749 Part 1 of removing log group, replacing with log set
- LOG-9749 Make CLI Favorites work with lists of log Ids in config
- LOG-9749 Add use of logsets ID in query and events requests

### CODE REVIEW

File: lecli/api_utils.py [hunk 1]
```diff
@@ -183,12 +182,12 @@ def get_account_resource_id():
...
     """
-    Get named log-key group from the config file.
+    Get named log-key list from the config file.
 
-    :param name: name of the group
+    :param name: name of the log key list
     """
 
-    section = 'LogGroups'
+    section = 'CLI_Favorites'
```
Comment (ilyabiryukov-r7) [inline lecli/api_utils.py]: Does this mean, users will need to update their configuration files?

Comment (pclune-r7) [reply lecli/api_utils.py]: yep - having the existing section LogGroup still in the config will not cause problems as it gets ignored - but they will need to add the new section for CLI_Favorites. On a fresh install the default config will have the new section and not the old. @stopal-r7 mentioned we will not be supporting backward compatibility on this.

Comment (ilyabiryukov-r7) [reply lecli/api_utils.py]: I wonder if we can include some code to automatically migrate users' configuration files for them. Should be simple enough, right?

File: lecli/query/commands.py [hunk 2]
```diff
@@ -11,14 +11,12 @@
 @click.command()
 # nargs (-1) makes sure multiple log keys is supported
 @click.argument('logkeys', type=click.STRING, nargs=-1)
-@click.option('-n', '--lognick', default=None,
-              help='Nickname of log in config file')
-@click.option('-g', '--loggroup', default=None,
-              help='Name of log group defined in config file')
+@click.option('-c', '--favorites', default=None,
```
Comment (pquinn-r7) [inline lecli/query/commands.py:14]: -c? I know -f is gone but this is not very intuitive.

File: lecli/api_utils.py (hunk 1 above)
Comment (pquinn-r7) [reply lecli/api_utils.py]: Yeah either rename if section found. OR read from both sections (CLI_favorites and LogGroups) as format is unchanged.

Comment (pclune-r7) [reply lecli/api_utils.py]: That's a change that'll need to be discussed with Product - I have not discussed the role out of these changes as Safa seemed pretty sure of what was to be done, but I'l follow up on it.

File: lecli/query/commands.py (hunk 2 above)
Comment (pclune-r7) [reply lecli/query/commands.py:14]: yea running out of letters, we're not going to get the most intuative - although in the case (at a reach ) 'c' for CLI Favorites?????

File: lecli/api_utils.py (hunk 1 above)
Comment (pclune-r7) [reply lecli/api_utils.py]: hhmmm If we are going to handle existing sections, I'm thinking maybe going with renaming the section as the original idea of this work is to move away from use of the terminology of LogGroups?

**Comment by pclune-r7** [comment]
Okay - gone with the following - grab all the items in the old section if old section exists, add them to new section, delete the old section.

File: lecli/api_utils.py [hunk 3]
```diff
@@ -91,6 +92,25 @@ def load_config():
         print_config_error_and_exit()
     if not CONFIG.has_section(AUTH_SECTION):
         print_config_error_and_exit(section=AUTH_SECTION)
+    if CONFIG.has_section(LOGGROUPS_SECTION):
+        remove_loggroup_section()
```
Comment (pquinn-r7) [inline lecli/api_utils.py]: replace_loggroup_section?

File: lecli/query/commands.py (hunk 2 above)
Comment (pquinn-r7) [reply lecli/query/commands.py:14]: Ok understood.

File: lecli/api_utils.py [hunk 4]
```diff
@@ -91,6 +92,25 @@ def load_config():
...
+
+
+def remove_loggroup_section():
```
Comment (pquinn-r7) [inline lecli/api_utils.py]: Can we get a test for this as well?

File: lecli/api_utils.py (hunk 1 above)
Comment (pquinn-r7) [reply lecli/api_utils.py]: Ok with me.

File: tests/test_apiutils.py [hunk 5]
```diff
@@ -230,6 +203,18 @@ def test_get_invalid_named_group_key(capsys):
             assert 'was not found' in out
 
 
+@patch('ConfigParser.ConfigParser')
+def test_replace_loggroup_section(mocked_configparser_class):
```
Comment (pclune-r7) [inline tests/test_apiutils.py:208]: @ilyabiryukov-r7 
Hi - this unit test was giving problems but I then got it passing - just not 100% if its solid - a review would be great - Thanks

**Comment by pclune-r7** [comment]
@fk92r7 
Hi Fergal - 
as Phil is not around to finish his review, appreciate you standing in! ; )
cheers

File: README.md [hunk 6]
```diff
@@ -71,21 +71,21 @@ The event and query functionality of the CLI supports a number of different ways
...
 
 Example usage: 
 ```
 lecli get recentevents 12345678-aaaa-bbbb-1234-1234cb123456 -l 200
 lecli get recentevents 12345678-aaaa-bbbb-1234-1234cb123456 -r 'last 2 hours'
-lecli get recentevents -n mynicknamedlog -l 200
-lecli get recentevents -g myloggroup -l 200
-lecli get recentevents -g myloggroup -r 'last 1 week'
+lecli get recentevents -c mylogalias -l 200
+lecli get recentevents --logset 12345678-aaaa-bbbb-1234-1234cb123457 -l 200
```
Comment (udangel-r7) [inline README.md:83]: Would it be possible to support logset by name? Seems very unintuitive to have to get the uuid.

File: README.md [hunk 7]
```diff
@@ -95,31 +95,24 @@ Example usage:
 lecli get events 12345678-aaaa-bbbb-1234-1234cb123456 -f 1465370400 -t 1465370500
 lecli get events 12345678-aaaa-bbbb-1234-1234cb123456 --datefrom '2016-05-18 11:04:00' --dateto '2016-05-18 11:09:59'
 lecli get events 12345678-aaaa-bbbb-1234-1234cb123456 -r 'yesterday'
-lecli get events --loggroup myloggroup --datefrom '2016-05-18 11:04:00' --dateto '2016-05-18 11:09:59'
-lecli get events --lognick mynicknamedlog --datefrom '2016-05-18 11:04:00' --dateto '2016-05-18 11:09:59'
-lecli get events --lognick mynicknamedlog -r 'last 3 weeks'
+lecli get events --logset 12345678-aaaa-bbbb-1234-1234cb123457 --datefrom '2016-05-18 11:04:00' --dateto '2016-05-18 11:09:59'
```
Comment (udangel-r7) [inline README.md:98]: I don't think we refer to logset as uuids in the ui

File: README.md [hunk 8]
```diff
@@ -162,30 +154,24 @@ First get the saved query id to run with `get savedqueries` command, then use th
...
 --------------------------------------------------
-The CLI supports the use of log nicknames and log groups via the configuration file. This makes searching well known or large lists of logs much simpler as you do not need to pass in lists of log Ids.
-
-####Log Nicknames
-Log nicknames allow an alias for a single log to be configured, this is done in the LogNicknames section of the configuration file. 
-```
-[LogNicknames]
-testlog = 12345678-aaaa-bbbb-1234-1234cb123456
-```
+The CLI supports the use in query commands of CLI Favorites (log alias) via the configuration file and log sets from the Logentries account. This makes searching well known or large lists of logs much simpler as you do not need to pass in lists of log Ids.
```
Comment (udangel-r7) [inline README.md]: It would probably be good to start with something like lecli supports command line favorites (CLI favorites) for query commands.

File: lecli/api_utils.py [hunk 9]
```diff
@@ -19,6 +19,8 @@
 
 AUTH_SECTION = 'Auth'
 URL_SECTION = 'Url'
+LOGGROUPS_SECTION = 'LogGroups'
+CLI_FAVORITES_SECTION = 'CLI_Favorites'
```
Comment (udangel-r7) [inline lecli/api_utils.py]: Everything else is camelcase/one word. Why the change? Or if you want to use underscore log groups should also have an underscore (prefer that one)

Comment (udangel-r7) [reply lecli/api_utils.py]: Also all others start with only one uppercase letter

File: README.md (hunk 6 above)
Comment (pclune-r7) [reply README.md:83]: I discuss this with Eoin and we agreed to do it this way, based on the following....

In the existing logset management api on lecli the user already uses logset ID for deletion, renaming and to get a specific logset; also it is more safer to use an ID as logset names on the web app are not  case insensitive meaning the wrong one could be used; also for implementation this way requires less requests and can use existing code.

File: README.md (hunk 7 above)
Comment (pclune-r7) [reply README.md:98]: see above comment

File: README.md (hunk 8 above)
Comment (pclune-r7) [reply README.md]: going to modify this slightly

File: lecli/api_utils.py (hunk 9 above)
Comment (pclune-r7) [reply lecli/api_utils.py]: I'm changing CLI_Favorites to Cli_Favorites. The LogGroups is there to be used with legacy section in config files users already have on their hosts and so cannot be changed here.

File: README.md (hunk 7 above)
Comment (ilyabiryukov-r7) [reply README.md:98]: @udangel-r7 what do you mean?

Comment (udangel-r7) [reply README.md:98]: Anything in the UI/documentation typically refers to logset as names. In insightops I think it is currently impossible (?) To get a logset uuid, the same case is I think in le with redisgns etc. 

I think we should at least explain how to get a logset key or explain that this is a unique key to identify a logset

**Review by udangel-r7 — APPROVED** [review]

File: README.md (hunk 7 above)
Comment (pclune-r7) [reply README.md:98]: @udangel-r7 
I reviewed the issue on logset uuid with Eoin - we want to stay with the use of logset uuid; but you were right about user access to the logset uuid as it is missing from the new webpage design - this is an oversight that Eoin is raising stories on to correct.
